File: src/Binary_Tree_Postorder_Traversal.py
# -*- coding: utf-8 -*-


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    # @param {TreeNode} root
    # @return {integer[]}
    def postorderTraversal(self, root):
        if not root:
            return []
        st, ret = [root], []
        vis = {}
        while len(st) != 0:
            node = st.pop()
            if str(node) in vis:
                ret.append(node.val)
            else:
                vis[str(node)] = True
                st.append(node)
                if node.right:
                    st.append(node.right)
                if node.left:
                    st.append(node.left)
        return ret


if __name__ == "__main__":
    a, b, c = TreeNode(1), TreeNode(2), TreeNode(3)
    a.left, a.right = b, c
    solution = Solution()
    print solution.postorderTraversal(a)

File: src/algorithms/cpp/Binary_Tree_Postorder_Traversal.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
#include <stack>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector <int> result;
        stack <TreeNode *> st, st2;
        map <TreeNode *, bool> mp;
        mp[NULL] = true;
        if (root)       st.push(root);
        while (!st.empty()) {
            TreeNode *node = st.top();
            st2.push(node), st.pop();
            if (node->right)    st.push(node->right);
            if (node->left)     st.push(node->left);
            while (!st2.empty() && mp[st2.top()->left] && mp[st2.top()->right]) {
                result.push_back(st2.top()->val);
                mp[st2.top()] = true;
                st2.pop();
            }
        }
        return result;
    }
};

int main() {
    return 0;
}
